c10/core/impl/LocalDispatchKeySet.h

```diff
@@ -24,57 +27,44 @@
 namespace c10 {
 namespace impl {
 
-// POD version of LocalDispatchKeySet.  Declared here just so that
-// we can put it in the guards.
-// This struct encapsulates special handling for TLS initialization
-// in set_included()/included() API so that they reflect the truth.
-// If you want to create PODLocalDispatchKeySet with non-zero state,
-// use set_included() instead of default constructor.
-struct C10_API PODLocalDispatchKeySet {
-  uint64_t included_;
-  uint64_t excluded_;
+// Helper class to handle conversion between the TLS zero initialized
+// convention and the "normal" in memory layout.
+class C10_API LocalDispatchKeySetWrapper {
+ public:
+  explicit LocalDispatchKeySetWrapper(PODLocalState* tls) : tls_(tls) {}
 
   // See Note [TLS Initialization]
   DispatchKeySet included() const {
-    return DispatchKeySet(DispatchKeySet::RAW, included_) ^
+    return DispatchKeySet(DispatchKeySet::RAW, tls_->included_) ^
         c10::default_included_set;
   }
   DispatchKeySet excluded() const {
-    return DispatchKeySet(DispatchKeySet::RAW, excluded_) ^
+    return DispatchKeySet(DispatchKeySet::RAW, tls_->excluded_) ^
         c10::default_excluded_set;
   }
 
   void set_included(DispatchKeySet x) {
-    included_ = (x ^ c10::default_included_set).raw_repr();
+    tls_->included_ = (x ^ c10::default_included_set).raw_repr();
   }
   void set_excluded(DispatchKeySet x) {
-    excluded_ = (x ^ c10::default_excluded_set).raw_repr();
+    tls_->excluded_ = (x ^ c10::default_excluded_set).raw_repr();
   }
+
+ private:
+  PODLocalState* tls_;
 };
-static_assert(
-    std::is_pod<PODLocalDispatchKeySet>::value,
-    "PODLocalDispatchKeySet must be a POD type.");
 
 struct C10_API LocalDispatchKeySet {
-  /* implicit */ LocalDispatchKeySet(PODLocalDispatchKeySet x)
-      : included_(x.included()), excluded_(x.excluded()) {}
+  LocalDispatchKeySet(LocalDispatchKeySetWrapper x) :
+    included_(x.included()),
+    excluded_(x.excluded()) {}
   DispatchKeySet included_;
   DispatchKeySet excluded_;
 };
 
-// thread_local variables cannot be C10_API on Windows.
-// Inlining this seems to break AutoDispatchBelowAutograd on Android.
-#if defined(_MSC_VER) || defined(C10_ANDROID)
-C10_API LocalDispatchKeySet tls_local_dispatch_key_set();
-#else // defined(_MSC_VER) || defined(C10_ANDROID)
-extern C10_API thread_local PODLocalDispatchKeySet raw_local_dispatch_key_set;
-
-inline C10_API LocalDispatchKeySet tls_local_dispatch_key_set() {
-  // Don't let people fiddle with the thread_local directly just
-  // because they include this header.
-  return raw_local_dispatch_key_set;
+inline C10_API LocalDispatchKeySet snapshot_tls_keyset() {
```

Comment:
I wonder what the "snapshot" here is supposed to mean. (Since you've gone through the trouble of renaming it I assume there's some motive here, but I can't figure it out from the code.) In particular, "snapshot" sounds like a verb but this function has no side effects.

Reply:
I guess you intend this to mean, "take a snapshot of the current tls keyset"? But that's a weird way to say it; if I define a getter for an integer field foobar on a class that I'm about to change, I don't call it `snapshot_foobar`, I just say `foobar` or `get_foobar`...

Reply:
The key point that I'm trying to convey is that this copies the data in the TLS rather than returning a view. So mutating the LocalDispatchKeySet that is returned does not propagate to the source state. But if that is confusing things more than helping I'm happy to go back to `tls_local_dispatch_key_set()`.

Reply:
I think `tls_keyset_snapshot` would do it. Also though, re "mutating does not propagate", why not just make LocalDispatchKeySet members immutable? That would take some of the weight off a name change - `tls_keyset_snapshot` is arguably more descriptive, but the diff shed some jitter if we kept the current one. 

Aside: I'm hopping around these PRs thinking about your keep-or-ditch question, so I realize this comment won't necessarily be useful - but just in case 😁

Reply:
I experimented with making the members `const`, but it broke certain builds unfortunately. (It wasn't clear what was going on so I didn't pursue it further.)

Reply:
In regular C++, if I have a function that returns a struct by value, and the struct is not obviously named like a reference (e.g., `ArrayRef` or `StringView`), then most reasonable readers will assume that value semantics apply here (in the same way that `std::vector<int> f()` gives you a fresh copy of the vector, it never aliases with anything else).

PyTorch blurs the lines here by defining `Tensor` to be a reference type. But in that case, `clone_tls_keyset()` is a good, idiomatic naming for this codebase, since clone is the name of the corresponding operation on tensors that doesn't give you an alias but in fact a new copy. I still maintain snapshot is totally unclear.

Reply:
Personally I don't think its meaning is unclear, but agreed that it's not a modifier in common use in the codebase. TBH I think the old name is fine, and would avoid some churn.

Reply:
> I experimented with making the members `const`, but it broke certain builds unfortunately. (It wasn't clear what was going on so I didn't pursue it further.)

It's probably worth a quick check of the errors to see if they're "real" in the sense that people are actually depending on the stateful updating of `LocalDispatchKeySet`'s properties, as opposed to just doing after-construction initialization, e.g. see my comment on the code in `InferenceMode`'s ctor.